core/trino-main/src/main/java/io/trino/operator/output/PagePartitioner.java

```diff
@@ -146,11 +148,51 @@ public void partitionPage(Page page, OperatorContext operatorContext)
         else {
             partitionPageByColumn(page);
         }
-        long positionsAppendersSizeAfter = getPositionsAppendersSizeInBytes();
-        flushPositionsAppenders(false);
+        long outputSizeInBytes = flushPositionsAppenders(false);
         updateMemoryUsage();
 
-        operatorContext.recordOutput(positionsAppendersSizeAfter - positionsAppendersSizeBefore, outputPositionCount);
+        recordOutputStats(operatorContext, outputSizeInBytes, outputPositionCount);
+    }
+
+    private void recordOutputStats(OperatorContext operatorContext, long outputSizeInBytes, int outputPositions)
+    {
+        // The outputSizeInBytes already reported during release
+        if (outputSizeReportedBeforeRelease > 0) {
+            long consumeFromPreviouslyReported = min(outputSizeReportedBeforeRelease, outputSizeInBytes);
+            outputSizeReportedBeforeRelease -= consumeFromPreviouslyReported;
+            outputSizeInBytes -= consumeFromPreviouslyReported;
+        }
+        if (outputSizeInBytes != 0 || outputPositions != 0) {
+            operatorContext.recordOutput(outputSizeInBytes, outputPositions);
+        }
+    }
+
+    /**
+     * Prepares this {@link PagePartitioner} for release to the pool by checking for dictionary mode appenders and either flattening
+     * them into direct appenders or forcing their current pages to flush to preserve a valuable dictionary encoded representation. This
+     * is done before release because we know that after reuse, the appenders will not observe any more inputs using the same dictionary.
+     */
+    public void prepareForRelease(OperatorContext operatorContext)
+    {
+        long bufferedSizeInBytes = 0;
+        long outputSizeInBytes = 0;
+        for (int partition = 0; partition < positionsAppenders.length; partition++) {
+            PositionsAppenderPageBuilder positionsAppender = positionsAppenders[partition];
+            Optional<Page> flushedPage = positionsAppender.flushOrFlattenBeforeRelease();
+            if (flushedPage.isPresent()) {
+                Page page = flushedPage.get();
+                outputSizeInBytes += page.getSizeInBytes();
+                enqueuePage(page, partition);
+            }
+            else {
+                // Dictionaries have now been flattened, so the new reported size is trustworthy to report
+                // eagerly
+                bufferedSizeInBytes += positionsAppender.getSizeInBytes();
+            }
+        }
+        updateMemoryUsage();
+        recordOutputStats(operatorContext, outputSizeInBytes + bufferedSizeInBytes, 0 /* no new positions */);
+        outputSizeReportedBeforeRelease += bufferedSizeInBytes;
```

Comment:
Do we actually need `outputSizeReportedBeforeRelease` as output page (flattened or not) is going to be produced one way or the other eventually? Why can't we just track `page.getSizeInBytes()` when the page is finally enqueued?

I feel like on one hand we are accounting size from produced pages, but also try to accommodate buffered pages somehow into that

With that in mind I feel like:
```
before = getPositionsAppendersSizeInBytes();
doSomething()
after = getPositionsAppendersSizeInBytes()
long outputSize = after - before
```
was easier to grasp

Reply:
> Why can't we just track page.getSizeInBytes() when the page is finally enqueued?

Because the final page flush / enqueue could occur outside of the context of an operator to report the size against (as a result of `PagePartitionerPool` and reuse). That was the reason why the logic was previously changed to eagerly report buffered bytes before flushing, but critically- that approach would incorrectly report output size for all dictionary encoded buffers because the size they report during `getPositionsAppendersSizeInBytes()` is not the actual size.

The previous logic may have seemed simpler, but was incorrect hence the need to fix it. The new logic can be summarized as:
1. Report `outputSizeInBytes` from produced `Page#getSizeInBytes()` when pages are flushed, just like was previously done before partitioner reuse and just like `TaskOutputOperator` does.
2. When releasing a partitioner for reuse (last chance where we are guaranteed to have an `OperatorContext` to report against) we report `Page#getSizeInBytes()` for any flushed pages, flatten any dictionary encoded appenders that didn't flush, and then report the appender buffered size in bytes eagerly (in case we don't have another chance to report that size in the future because the partitioner is _not_ reused). Since we no longer have any dictionary encoded appenders at this point, we know that the reported buffer size accurately predicts the size of the page that would be flushed if no additional insertions occur which was not before. 
3. In the case where the partitioner _is_ reused, and new positions _are_ appended- we have already accounted for the buffered size on the last release. We need to subtract that amount before reporting any additional output bytes after reuse.